### PR TITLE
Replace 'es2015-i18n-tag' w/ 'lingui'

### LINKED ISSUE #85: Change i18n tooling to [`lingui`](https://lingui.dev)
## Pros

- _Little_ bit bigger footprint than the one currently used ([`es2015-i18n-tag`](https://github.com/skolmer/es2015-i18n-tag)) ([3.4kb](https://bundlephobia.com/package/@lingui/core@3.17.0)vs [2.4kb](https://bundlephobia.com/package/es2015-i18n-tag@1.6.1)).
- Works with [SWC](https://lingui.dev/ref/macro.html#swc) _and_ [Babel](https://lingui.dev/ref/macro.html#babel). Gives us compatibility for Storybook.
- Possibility to [generate catalog files based upon i18n usage in component files](https://lingui.dev/tutorials/cli#extracting-messages).
- [Validation of translation coverage](https://lingui.dev/tutorials/cli#validation-of-message-catalogs)

## Cons

- A bit unsure if it supports the JSON structure `create-ignite-app` is using today. May be possible through some configuration. Though, they have a [pretty nice alternative](https://lingui.dev/ref/catalog-formats#json), where you have one translation files [per component](https://lingui.dev/ref/conf#separate-catalogs-per-component-placed-inside-shared-directory). Though, these are just examples. We should strive for keeping the current pattern and then reconsolidate if another one would be better.

### PR DESCRIPTION
**Pull request by adamsoderstrom**
As the summary says, this PR replaces `es2015-i18n-tag` tooling with `lingui`.

`lingui` comes with a few nice tools and workflows, but also results in SWC support for the Next application in the repo.

**note:** this PR doesn't include migration to SWC. `lingui` supports it, though.

### CODE REVIEW

File: src/pages/_app.tsx [hunk 1]
```diff
@@ -38,12 +37,7 @@ function App(props: AppProps) {
         />
       </Head>
 
-      <RootProvider
-        emotionCache={emotionCache}
-        defaultLocale={defaultLocale}
-        locale={locale}
-        {...other}
-      >
+      <RootProvider emotionCache={emotionCache} locale={locale} {...other}>
```
Comment (maeertin) [inline src/pages/_app.tsx:40]: So `defaultLocale` is no longer needed? :o

**Review by maeertin — CHANGES_REQUESTED** [review]
Sooooo awesome to get this kick-started, great job @adamsoderstrom 🔥✨😄 One thing to consider though is if this PR should contain the whole migration to SWC or if it should only implement lingui. Let's discuss this 👍

File: .eslintignore [hunk 2]
```diff
@@ -1,5 +1,6 @@
 /.git
 /coverage
+/public/locales/*/*.js
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply .eslintignore:3]: Ah, that sounds reasonable.

Will try it out!

File: .storybook/main.ts [hunk 3]
```diff
@@ -2,8 +2,54 @@ import { merge } from 'webpack-merge'
 import type { StorybookConfig } from '@storybook/nextjs'
 import webpackBaseConfig from '../webpackBaseConfig'
 
+const productionPlugins = ['@babel/plugin-transform-react-constant-elements']
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply .storybook/main.ts:5]: Ah, alright!

Will remove!

File: .storybook/main.ts [hunk 4]
```diff
@@ -2,8 +2,54 @@ import { merge } from 'webpack-merge'
...
+        'next/babel',
+        {
+          // Target ES2015+ browsers
+          'preset-env': {
+            targets: 'Chrome >= 60, Safari >= 10.1, iOS >= 10.3, Firefox >= 54, Edge >= 15',
+            useBuiltIns: false,
+          },
+        },
+      ],
+    ],
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply .storybook/main.ts:21]: Ah, true! Will try out removing it

File: .storybook/preview.js [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import '../scripts/globals'
 import * as React from 'react'
 import { useColorScheme } from '@mui/material'
+import { messages as i18n } from '@lingui/loader!../public/locales/en/messages.po'
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply .storybook/preview.js:4]: Ah, yeah. Will try it out

File: src/layouts/AppBase/partials/AppBaseMarketDialog.js [hunk 6]
```diff
@@ -80,15 +77,14 @@ const AppBaseMarketDialog = React.memo(function AppBaseMarketDialog(props) {
 
       <DialogContent>
         <Typography variant="body2" paragraph>
-          {t(
-            __translationGroup,
-          )`Please select the country to which your order will be sent. This will give you accurate pricing, delivery times and shipping costs for your destination.`}
+          Please select the country to which your order will be sent. This will give you accurate
+          pricing, delivery times and shipping costs for your destination.
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply src/layouts/AppBase/partials/AppBaseMarketDialog.js:81]: Ah, of course! Thanks!

File: src/pages/_app.tsx (hunk 1 above)
Comment (adamsoderstrom) [reply src/pages/_app.tsx:40]: Very good question.
We'll probably need it for monolingual apps?

Will try it out

File: package.json [hunk 7]
```diff
@@ -29,7 +29,9 @@
     "storybook": "storybook dev -p 3001 --ci",
     "storybook:build": "storybook build -o .storybook-static",
     "storybook:test": "test-storybook --url http://127.0.0.1:3001",
-    "typecheck": "tsc --noEmit"
+    "typecheck": "tsc --noEmit",
+    "extract": "lingui extract",
+    "compile": "lingui compile"
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply package.json:34]: Ah, yeah. Thinking of `yarn lingui:extract` and `yarn lingui:compile`.

What do you say?

File: package.json [hunk 8]
```diff
@@ -60,10 +63,15 @@
...
     "@babel/preset-typescript": "^7.18.6",
+    "@babel/preset-react": "^7.18.6",
     "@commitlint/cli": "^17.0.2",
     "@commitlint/config-conventional": "^17.0.2",
     "@commitlint/cz-commitlint": "^17.0.0",
     "@emotion/babel-plugin": "^11.9.2",
+    "@lingui/cli": "^3.17.2",
+    "@lingui/loader": "^3.17.2",
+    "@lingui/macro": "^3.17.2",
+    "@lingui/swc-plugin": "^0.2.3",
```
Comment (adamsoderstrom) [inline package.json:74]: This should also be removed!

File: public/locales/en/messages.po [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+"Report-Msgid-Bugs-To: \n"
+"PO-Revision-Date: \n"
+"Last-Translator: \n"
+"Language-Team: \n"
+"Plural-Forms: \n"
+
+#: src/containers/CartItem/CartItem.js:108
+msgid "Decrease quantity to {0}"
+msgstr ""
+
```
Comment (adamsoderstrom) [inline public/locales/en/messages.po:19]: This file should probably also be removed.
Maybe on `npx @noaignite/create-app`?

Comment (maeertin) [reply public/locales/en/messages.po:19]: So you mean this file does not need to exist for the app to work?

File: package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (maeertin) [reply package.json:34]: Sounds good @adamsoderstrom !

File: .storybook/main.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply .storybook/main.ts]: Since i've now reverted the changes in terms of babel, the `preset` array is still present, because Next.js needs them(?)

File: next.config.js [hunk 10]
```diff
@@ -12,7 +10,9 @@ const nextConfig = {
     locales: ['en'],
   },
   poweredByHeader: false,
-  webpack: (config) => merge(config, webpackBaseConfig),
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply next.config.js:15]: I've now reverted the removal of the `webpack` property (since we use babel for Next again).

File: package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply package.json]: I've now done the above.

File: .storybook/preview.js [hunk 11]
```diff
@@ -62,7 +63,7 @@ function ModeSwitcher(props) {
 
 export const decorators = [
   (Story, context) => (
-    <RootProvider settings={settings}>
+    <RootProvider settings={settings} i18n={i18n} locale="en">
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply .storybook/preview.js]: I've now removed the `i18n={i18n}` statement. The import statement as well.

File: src/api/i18n.js [hunk 12]
```diff
@@ -0,0 +1,7 @@
+async function loadTranslations(locale) {
+  const { messages } = await import(`@lingui/loader!../../public/locales/${locale}/messages.po`)
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply src/api/i18n.js]: Very good question.
I've now changed it to the previous try/catch pattern without actually investigating if it's needed.

File: src/api/i18n.js [hunk 13]
```diff
@@ -0,0 +1,7 @@
+async function loadTranslations(locale) {
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply src/api/i18n.js]: Ah, totally agree.
I've removed this file now, so we don't need to take any further actions on this.

File: src/api/index.js [hunk 14]
```diff
@@ -1 +1,2 @@
 export * from './utils'
+export { default as loadCatalog } from './i18n'
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply src/api/index.js]: Ah, yeah, totally. My bad. File removed, so this is now resolved.

File: package.json [hunk 15]
```diff
@@ -60,10 +63,15 @@
     "@babel/core": "^7.14.6",
     "@babel/plugin-transform-react-constant-elements": "^7.10.1",
```
[the remark this answers is not in the document]
Comment (adamsoderstrom) [reply package.json:64]: Agree, but since we discussed that we will use Babel for Next.js even after this PR, it will still be necessary.

**Comment by adamsoderstrom** [comment]
One thing that we still need to do before merging this, is to ensure that we provide a compiled translation file in every production build.

The production builds are currently gitignore'd in this PR (`/public/locales/*/*.js`). I wonder if we shall alter the `yarn build` script to account for this? @maeertin

**Comment by adamsoderstrom** [comment]
> One thing that we still need to do before merging this, is to ensure that we provide a compiled translation file in every production build.
> 
> The production builds are currently gitignore'd in this PR (`/public/locales/*/*.js`). I wonder if we shall alter the `yarn build` script to account for this? @maeertin

I think we can skip this, tbh.

[In `next.js`'s own example of using `lingui`](https://github.com/vercel/next.js/tree/canary/examples/with-lingui#with-lingui-example), they state:

> It adds a webpack loader for the messages to avoid having to manually compile while developing as well as adds the compile step to the next build script for production builds.

Webpack loader, in this case, is `@lingui/loader`, i assume.

So we should be fine now! 😊

File: src/contexts/Root/index.js [hunk 16]
```diff
@@ -19,14 +19,14 @@ function RootProvider(props) {
...
-    locale,
     initialSelection,
+    locale,
     settings,
   } = props
 
   return (
     <CacheProvider value={emotionCache}>
-      <I18nProvider defaultLocale={defaultLocale} locale={locale}>
+      <I18nProvider locale={locale} defaultLocale={defaultLocale}>
```
Comment (maeertin) [inline src/contexts/Root/index.js:29]: This feels like an unnecessary change? 😯

File: src/layouts/App/partials/AppCartDrawer.js [hunk 17]
```diff
@@ -61,7 +61,7 @@ const AppCartDrawer = React.memo(function AppCartDrawer(props) {
 
       <div>
         <Button component={RouterLink} href="/checkout" variant="contained" fullWidth>
-          {t(__translationGroup)`To checkout`}
+          {t` To checkout`}
```
Comment (maeertin) [inline src/layouts/App/partials/AppCartDrawer.js:64]: There's an incorrect whitespace here ;)

File: src/pages/_app.tsx (hunk 1 above)
Comment (maeertin) [reply src/pages/_app.tsx:40]: What do you mean? I can't see that `defaultLocale` has been re-added to this file. It needs to be added here for the `defaultLocale` set in `next.config.js` to be forwarded to the I18nProvider.

**Review by maeertin — CHANGES_REQUESTED** [review]
Nice further updates, added some more comments :)

**Review by maeertin — APPROVED** [review]
I helped out with some of the comments that I gave from the previous review so I think everything now looks good in this PR 😄 I also noticed Storybook was crashing due a js error regarding i18n so I fixed it.

Is there any more testing that we might want to do before closing and mergin this PR? 🚀 

Awesome job @adamsoderstrom ✨

**Comment by adamsoderstrom** [comment]
Ah, hahah! Sweet, @maeertin !
Started working on your comments from today locally, but never got to pushing them! 😄

**Comment by adamsoderstrom** [comment]
> Is there any more testing that we might want to do before closing and mergin this PR? 🚀

Yeah. I'll add some comments to the PR and then someone (or me) can have a look on them! 😊
@maeertin

File: lingui.config.js [hunk 18]
```diff
@@ -0,0 +1,15 @@
+/** @type {import('@lingui/conf').LinguiConfig} */
+module.exports = {
+  locales: ['en'],
+  extractBabelOptions: {
+    presets: ['@babel/preset-typescript', '@babel/preset-react'],
+  },
+  sourceLocale: 'en',
```
Comment (adamsoderstrom) [inline lingui.config.js:7]: We should source these from the `next.config.js` file.

```suggestion
  locales: nextConfig.i18n.locales,
  extractBabelOptions: {
    presets: ['@babel/preset-typescript', '@babel/preset-react'],
  },
  sourceLocale: nextConfig.i18n.defaultLocale,
```

File: src/contexts/I18n/index.js [hunk 19]
```diff
@@ -13,35 +14,25 @@ export function useI18n() {
...
-      defaultLocale,
-      locale,
-      t: i18n,
-    }),
-    [defaultLocale, locale],
-  )
-
-  return <I18nContext.Provider value={contextValue}>{children}</I18nContext.Provider>
+  i18n.load(locale, translations)
+  i18n.activate(locale)
```
Comment (adamsoderstrom) [inline src/contexts/I18n/index.js:33]: I'd say that this solution *works*. Maybe we can wrap it in a `useEffect`? 

My biggest question is regarding what impact it would have by wrapping it in a `useEffect` [as in `next.js/examples/with-lingui`](https://github.com/vercel/next.js/blob/2abc824869e4927b820767c1c416a381c8ae528d/examples/with-lingui/pages/_app.js#L15-L24)

**Review by adamsoderstrom — COMMENTED** [review]
Added two comments.
One that i think we should act upon before merging (sourcing locales from `next.config.js` in `lingui.config.js`). The other one is not that important, i'd say.

File: lingui.config.js (hunk 18 above)
Comment (maeertin) [reply lingui.config.js]: Totally agree, in a lot of projects we add a `locales.js` file in the root of the project that is then used by `next.config.js` and a potential Sanity setup. This file could then be used for `lingui.config.js` as well :)

File: src/contexts/I18n/index.js (hunk 19 above)
Comment (maeertin) [reply src/contexts/I18n/index.js:33]: Let's try it out when we have the chance :)

File: lingui.config.js (hunk 18 above)
Comment (adamsoderstrom) [reply lingui.config.js]: Ah, yeah. I've implemented this now! @maeertin

**Review by alexanderflink — APPROVED** [review]
This looks great! Thanks @adamsoderstrom 😀

**Review by ghost — APPROVED** [review]

**Comment by maeertin** [comment]
You can now run `npx @noaignite/create-app my-app --branch feature/lingui-swc` to test it out in a fresh install :)

In the case that it doesn't work, you might need to delete the npx cache found at `~/.npm/_npx`.

@adamsoderstrom

**Comment by adamsoderstrom** [comment]
> You can now run npx @noaignite/create-app my-app --branch feature/lingui-swc to test it out in a fresh install :)
>
> In the case that it doesn't work, you might need to delete the npx cache found at ~/.npm/_npx.

Worked good after removing `~/.npm/_npx`, @maeertin!
According to me, things looks good after building the application as well. 😊

**Comment by maeertin** [comment]
Are we ready for merge you think @adamsoderstrom? :)

**Comment by adamsoderstrom** [comment]
> Are we ready for merge you think @adamsoderstrom? :)

Yeah, i guess! 😊 @maeertin

**Comment by maeertin** [comment]
Actually what's missing in this PR is to entirely clean up the previous solution. Don't know what remains might be in the project from it but we should at least remove the old npm packages `es2015-i18n-tag` and `babel-plugin-i18n-tag-translate`.

**Comment by adamsoderstrom** [comment]
> Actually what's missing in this PR is to entirely clean up the previous solution. Don't know what remains might be in the project from it but we should at least remove the old npm packages `es2015-i18n-tag` and `babel-plugin-i18n-tag-translate`.

True. Will remove `es2015-i18n-tag` and `babel-plugin-i18n-tag-translate`.
I've done some cleanup, but will look into more "leftovers"

@maeertin

**Comment by maeertin** [comment]
Also just found that `babel-plugin-i18n-tag-translate` is still used in `babel.config.js` :) @adamsoderstrom

**Comment by maeertin** [comment]
Let's also remove the old `I18nContext` and bump lingui to v4, which removes the need for `make-plural/plurals` as we discussed @adamsoderstrom

**Comment by adamsoderstrom** [comment]
There seems to be a JSX-related bug in the `4.0.0` version of Lingui. Created an issue (https://github.com/lingui/js-lingui/issues/1638)

**Comment by adamsoderstrom** [comment]
Seems like [it wasn't a bug](https://github.com/lingui/js-lingui/issues/1638#issuecomment-1537940643)!

I've now made a following commit (), which strives to follow @thekip's instructions.
May be some refactoring to be done (use default + custom extractor, maybe).

@maeertin

File: src/containers/CartItem/CartItem.js [hunk 20]
```diff
@@ -51,7 +52,6 @@ function CartItem(props) {
   const { product } = cartItem
 
   const { decreaseItem, increaseItem, removeItem } = useCentraHandlers()
-  const { t } = useI18n()
```
Comment (thekip) [inline src/containers/CartItem/CartItem.js:54]: You have to use useLingui hook if you are using `t` macro inside a react components. Otherwise component will not be updated when language is switched.  Or use <Trans> react macro, which will automatically subscribed for changes.  

So for smoother migration i will suggest to replace all `const { t } = useI18n()` to `useLingui()` (even if the result of it is not used)

File: src/contexts/I18n/index.js (hunk 19 above)
Comment (thekip) [reply src/contexts/I18n/index.js:33]: Example on nextjs side is broken. Please use examples from lingui repo. With example from nextjs repo you will never get SSR working.

**Comment by thekip** [comment]
Guys, I will add my 50 cents here.

1. the current state of your integration will not support Next's  SSR / SSG. I didn't see any `getServerSideProps` / `getStaticProps`. Check examples: https://github.com/lingui/js-lingui/tree/main/examples and working nextjs project here https://github.com/alt3/rank-my-wallet
2. I recommend to use `<Trans>Hello {name}</Trans>` whenever possible. `t` macro is confusing and easy to misuse. 
3. You also can leverage an automatic page level catalog splitting by using "[experimental extractor](https://lingui.dev/guides/message-extraction#dependency-tree-crawling-experimental)", here is an example https://github.com/alt3/rank-my-wallet/pull/148 you still will have issues with JSX inside JS files, because it uses esbuild under the hood, which most likely will not understand it.

**Comment by maeertin** [comment]
Closing PR as "cia" is now purely a documentation app.
